media-proxy/include/mesh/st2110rx.h

```diff
@@ -0,0 +1,176 @@
+#ifndef ST2110RX_H
+#define ST2110RX_H
+
+#include "st2110.h"
+
+namespace mesh::connection {
+
+/**
+ * ST2110Rx
+ *
+ * Base abstract class of ST2110Rx. ST2110_20Rx/ST2110_22Rx/ST2110_30Rx
+ * inherit this class.
+ */
+template <typename FRAME, typename HANDLE, typename OPS> class ST2110Rx : public ST2110 {
+  public:
+    ST2110Rx() : mtl_session(nullptr), ops{0}, transfer_size(0) { _kind = Kind::receiver; }
+
+    ~ST2110Rx() {
+        shutdown(_ctx);
+        if (ops.name)
+            free((void *)ops.name);
+    }
+
+  protected:
+    HANDLE mtl_session;
+    OPS ops;
+    size_t transfer_size;
+    std::jthread frame_thread_handle;
+    context::Context _ctx = context::WithCancel(context::Background());
+
+    virtual FRAME *get_frame(HANDLE) = 0;
+    virtual int put_frame(HANDLE, FRAME *) = 0;
+    virtual HANDLE create_session(mtl_handle, OPS *) = 0;
+    virtual int close_session(HANDLE) = 0;
+
+    int configure_common(context::Context& ctx, const std::string& dev_port,
```

Comment:
wait, why there is code in header files? Please remove all function definitions from .h and move them to .cc. leave just declarations.

Reply:
Well, this shouldn't be that strict rule. It's quite a typical and industry-wide adopted code style. Let Tomasz decide on that.

Reply:
I wont be moving that code to ```.cc``` file simply because template implementation have to be in header file. I can only move implementation to subsequent header ```st2110rx_impl.h```and include it everywhere.

Reply:
I believe, `st2110rx_impl.h` is not needed for now.

Reply:
I don't know. Placing code in a header file will increase compilation time because the code will be compiled in every file it is included in.

This could be handy for example for small libraries where all the code is in a header file, but I'm not sure if this is good design choose for our code-style.

Reply:
I belive the guards that we have in every header file do the job
```cpp
#ifndef HEADER_FILE_H
#define HEADER_FILE_H

// Header file content

#endif // HEADER_FILE_H
```

Reply:
This is a template class, it will be compiled only if you create subclass.
Also header it is protected by macro to avoid multiple compilation 
```
#ifndef ST2110RX_H
#define ST2110RX_H
#endif
```
I dont think this is an issue

Reply:
the #ifndef guard will ensure that it is included just ones in every translation unit, but this can be included in multiple translation unit, basically in multiple .cc files.

The argument about template class sounds convincing. I have to check it

Reply:
Yeah, the guards allow including a header file multiple times in `.cc` and `.h` files ensuring it will be compiled only once.

Reply:
if it was just normal executable code ( I'm still not sure about template class) it would be compiled multiple times. Every source file is preprocessed separately ( so every include in sticked into every source file ) the result of this is translation unit which is then compiled. If we add code into header file it will be copied to every source file at preprocessing and then compiled multiple times. 

`#define HEADER_FILE_H` works only in single translation unit it is not propagated to all, so `#ifndef HEADER_FILE_H` ensures that we won't include single header file multiple times to a single source file, but it has to be included to every source file that actually needs it, and it will.

Reply:
Oh, I didn't think about this from that perspective, so may be you're right. But still, there shouldn't be any significant latency introduced by this code style in this project. Let's keep it as it is.

Reply:
ok, linker is smart. The linker eliminates duplicate template instantiations is a result of ODR merging, implemented via mechanisms like weak symbols or COMDAT folding.

It was nice to discuss. I think this code in header is ok.